bin/kmstool-enclave-cli/main.c

```diff
@@ -40,71 +49,114 @@ struct app_ctx {
+    const struct aws_string *key_id;
+    enum aws_key_spec key_spec;
 };
 
-static void s_usage(int exit_code) {
-    fprintf(stderr, "usage: kmstool_enclave_cli [options]\n");
+/*
+ * Function to print the different commands
+ */
+static void print_commands(int exit_code) {
+    fprintf(stderr, "usage: kmstool_enclave_cli [command]\n");
+    fprintf(stderr, "\n Commands: \n\n");
+    fprintf(stderr, "    decrypt: Decrypt a given ciphertext blob.\n");
+    fprintf(stderr, "    genkey: Generate a datakey from KMS encrypted with the given key id.\n");
+    exit(exit_code);
+}
+
+/*
+ * Function to print out the argumetns for decrypt
+ */
+static void s_usage_decrypt(int exit_code) {
+    fprintf(stderr, "usage: kmstool_enclave_cli decrypt [options]\n");
     fprintf(stderr, "\n Options: \n\n");
-    fprintf(stderr, "    --region REGION: AWS region to use for KMS\n");
+    fprintf(stderr, "    --help: Displays this message and exits\n");
+    fprintf(stderr, "    --region REGION: AWS region to use for KMS. Default: 'us-east-1'\n");
     fprintf(stderr, "    --proxy-port PORT: Connect to KMS proxy on PORT. Default: 8000\n");
     fprintf(stderr, "    --aws-access-key-id ACCESS_KEY_ID: AWS access key ID\n");
     fprintf(stderr, "    --aws-secret-access-key SECRET_ACCESS_KEY: AWS secret access key\n");
     fprintf(stderr, "    --aws-session-token SESSION_TOKEN: Session token associated with the access key ID\n");
     fprintf(stderr, "    --ciphertext CIPHERTEXT: base64-encoded ciphertext that need to decrypt\n");
     fprintf(stderr, "    --key-id KEY_ID: decrypt key id (for symmetric keys, is optional)\n");
-    fprintf(
-        stderr,
-        "    --encryption-algorithm ENCRYPTION_ALGORITHM: encryption algorhthm for ciphertext (is required if key-id "
-        "exists)\n");
-    fprintf(stderr, "    --help: Display this message and exit\n");
+    fprintf(stderr, "    --encryption-algorithm ENCRYPTION_ALGORITHM: encryption algorithm for ciphertext\n");
+    exit(exit_code);
+}
+
+/*
+ * Function to print out the arguments for genkey
+ */
+static void s_usage_genkey(int exit_code) {
+    fprintf(stderr, "usage: kmstool_enclave_cli genkey [options]\n");
+    fprintf(stderr, "\n Options: \n\n");
+    fprintf(stderr, "    --help: Displays this message and exits\n");
+    fprintf(stderr, "    --region REGION: AWS region to use for KMS. Default: 'us-east-1'\n");
+    fprintf(stderr, "    --proxy-port PORT: Connect to KMS proxy on PORT. Default: 8000\n");
+    fprintf(stderr, "    --aws-access-key-id ACCESS_KEY_ID: AWS access key ID\n");
+    fprintf(stderr, "    --aws-secret-access-key SECRET_ACCESS_KEY: AWS secret access key\n");
+    fprintf(stderr, "    --aws-session-token SESSION_TOKEN: Session token associated with the access key ID\n");
+    fprintf(stderr, "    --key-id KEY_ID: key id\n");
+    fprintf(stderr, "    --key-spec KEY_SPEC: The key spec used to create the key (AES-256 or AES-128).\n");
```

Comment:
Could we add an option for `NumberOfBytes` to align with the [KMS API](https://docs.aws.amazon.com/kms/latest/APIReference/API_GenerateDataKey.html#KMS-GenerateDataKey-request-NumberOfBytes)? Users would have to specify `--key-spec` OR `num-bytes`.

Reply:
At the moment I don't believe that the underlying C SDK has implemented this parameter. While I agree that it would be a valuable enhancement, I don't think it should block merging this PR.